dbuf-lsp/src/navigation_handler/navigation/find_definition.rs

```diff
@@ -0,0 +1,177 @@
+//! Find definition of symbol
+//!
+
+use dbuf_core::ast::parsed::Pattern;
+use dbuf_core::ast::parsed::PatternNode;
+use dbuf_core::ast::parsed::TypeDefinition;
+use tower_lsp::lsp_types::Range;
+
+use crate::core::ast_access::ElaboratedHelper;
+use crate::core::ast_access::LocStringHelper;
+use crate::core::ast_access::LocationHelpers;
+use crate::core::ast_access::{Loc, Str};
+use crate::core::dbuf_language::get_bultin_types;
+
+use crate::core::navigator::Navigator;
+use crate::core::navigator::Symbol;
+
+pub fn find_definition_impl(navigator: &Navigator, symbol: &Symbol) -> Option<Range> {
+    match symbol {
+        Symbol::Type(t) => {
+            if get_bultin_types().contains(t) {
+                return None;
+            }
+            navigator
+                .parsed
+                .iter()
+                .find(|d| d.name.as_ref() == t)
+                .map(|d| d.name.get_location().to_lsp())
+                .unwrap_or_else(|| {
+                    panic!("type not found\n{:#?}", symbol);
+                })
+                .into()
+        }
+        Symbol::Dependency { t, dependency } => {
+            let dep = navigator
+                .parsed
+                .iter()
+                .find(|d| d.name.as_ref() == t)
+                .map(|d| &d.data.dependencies);
+            if let Some(dependencies) = dep {
+                dependencies
+                    .iter()
+                    .find(|d| d.name.as_ref() == dependency)
+                    .map(|d| d.name.get_location().to_lsp())
+                    .unwrap_or_else(|| {
+                        panic!("dependency not found\n{:#?}", symbol);
+                    })
+                    .into()
+            } else {
+                panic!("dependency not found\n{:#?}", symbol);
+            }
+        }
+        Symbol::Field { constructor, field } => {
+            let elaborated = navigator.elaborated;
+            let t = elaborated
+                .get_constructor_type(constructor)
+                .unwrap_or_else(|| {
+                    panic!("field not found\n{:#?}", symbol);
+                });
+            let body = navigator
+                .parsed
+                .iter()
+                .find(|d| d.name.as_ref() == t)
+                .map(|d| &d.data.body);
+
+            match body {
```

Comment:
`match` is an expression so you can use it to produce value to return instead of `return`ing explicitly